runtime/src/bank.rs

```diff
@@ -11305,4 +11314,53 @@ pub(crate) mod tests {
         }
         assert_eq!(bank.get_inflation_num_slots(), 2 * slots_per_epoch);
     }
+
+    #[test]
+    fn test_stake_vote_account_validity() {
+        let validator_vote_keypairs0 = ValidatorVoteKeypairs::new_rand();
+        let validator_vote_keypairs1 = ValidatorVoteKeypairs::new_rand();
+        let validator_keypairs = vec![&validator_vote_keypairs0, &validator_vote_keypairs1];
+        let GenesisConfigInfo {
+            genesis_config,
+            mint_keypair: _,
+            voting_keypair: _,
+        } = create_genesis_config_with_vote_accounts(
+            1_000_000_000,
+            &validator_keypairs,
+            vec![10_000; 2],
+        );
+        let bank = Arc::new(Bank::new(&genesis_config));
+        let stake_delegation_accounts = bank.stake_delegation_accounts(&mut null_tracer());
+        assert_eq!(stake_delegation_accounts.len(), 2);
+
+        // Modify staked vote account owner; a vote account owned by another program could be
+        // freely modified with malicious data
+        let bogus_vote_program = Pubkey::new_unique();
+        let mut vote_account = bank
+            .get_account(&validator_vote_keypairs0.vote_keypair.pubkey())
+            .unwrap_or_default();
+        vote_account.owner = bogus_vote_program;
+        bank.store_account(
```

Comment:
I think runtime enforces an intermediately write of account.lamports == 0 when chainging existing non-system accounts.
So, how about inserting this here?:

```rust
// simulate vote account removal via full withdrawal
bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &Acount::default());
```

Then, the following assertion should be `1`: 

```rust
assert_eq!(bank.vote_accounts().len(), 1);
```

Reply:
That is more realistic, but it doesn't change `bank.vote_accounts().len()`. (Confirmed with test run.)
Unless I'm missing something, I don't see that Stakes ever removes vote_accounts, regardless of owner or lamports balance.

Reply:
well, sorry, my comment was partly wrong.

Account::default().owner isn't Vote1111; that's why it didn't change.

So, the correct code should be like this:

```
vote_account.lamports = 0;
// simulate vote account removal via full withdrawal
bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &vote_account);

vote_account.lamports = original_lamports;
vote_account.owner = bogus_vote_program;

bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &vote_account);
assert_eq!(bank.vote_accounts().len(), 1);

```

Well, Stake::store is hard to read without much aid; Please check this out: https://github.com/solana-labs/solana/pull/14065/files#r540752554 I still think it should remove `vote_accounts`.

Reply:
Oh thank you, I did miss that all owned vote accounts are removed and then potentially re-added.
Okay, great, your suggested code looks just like mine does now :)